src/decisionengine/framework/engine/SourceWorkers.py

```
@@ -0,0 +1,191 @@
+            return self.use_count.value == 0
+
+    def increment_use_count(self):
+        with self.use_count.get_lock():
+            self.use_count.value += 1
+
+    def decrement_use_count(self):
+        with self.use_count.get_lock():
+            self.use_count.value -= 1
+
+    def take_offline(self):
+        with self.use_count.get_lock():
+            self.use_count.value = 0
+
+    def run(self):
+        """
+        Get the data from source
+        """
+        self.logger.info(f"Starting source loop for {self.key}")
+        SOURCE_ACQUIRE_GAUGE.labels(self.key)
+        with producers[self.connection].acquire(block=True) as producer:
+            # If task manager is in offline state, do not keep executing sources.
+            while not self.should_stop():
+                try:
+                    self.logger.info(f"Source {self.name} calling acquire")
+                    data = self.module_instance.acquire()
+                    Module.verify_products(self.module_instance, data)
+                    self.logger.info(f"Source {self.name} acquire returned")
+                    SOURCE_ACQUIRE_GAUGE.labels(self.name).set_to_current_time()
+                    self.logger.debug(f"Publishing data to queue {self.key} with routing key {self.key}")
+                    producer.publish(
+                        dict(source_module=self.module, class_name=self.name, data=data),
+                        routing_key=self.key,
+                        exchange=self.exchange,
+                        serializer="pickle",
+                        declare=[
+                            self.exchange,
+                            self.queue,
+                        ],
+                    )
+                    self.logger.info(f"Source {self.name} {self.module} finished cycle")
+                except Exception:
+                    self.logger.exception(f"Exception running source {self.name} ")
+                    producer.publish(
```

Did we want to re-raise the exception rather than republish or am I missing something here?

---

When an exception occurs while running the source, any downstream channels must be taken offline.  This is achieved by publishing a message where the data is `State.SHUTDOWN` (see next line).  Any listening channels consume that message and enter the shutting-down state.

The source's exception is logged above and the source loop terminates after publishing the `State.SHUTDOWN` message.

---

Ahh, I get it.  Somehow I didn't see `data=State.SHUTDOWN`.....